### ISSUE TITLE
Make updates endpoint blazingly fast

### ISSUE THREAD

**Issue by WcaleNieWolny** [Cap-go/capgo#398]
Currently the updates endpoint is implemented very inefficiently. This was attempted to be fixed by implementing [redis](https://github.com/Cap-go/capgo/blob/d8a70c896b7a44cb289bc0b0fc0de111030b894c/supabase/functions/_utils/redis.ts) but the solution was buggy at best and completely unreliable at worst.

I think this can be broken down to the following reasons:
 * Redis was never given access to PostgreSQL and it was implemented as a middleware
 * Data was invalidated from an edge function that also was not REALLY that reliable
 * There was a lack of IOS tests done prior to merging redis
 * The underling problem with the slow endpoint was never fixed
 * The entire chain (loadbalancer -> edge fn -> redis ->  supabase backend (supabase HTTP api) -> postgress) was kept
 
 Here is what currently happens on an update endpoint (with redis disabled):
 
  - First we trigger the cloudflare load balancer
  - Second we trigger a supabase edge function (or a different provider edge fn)
  - This edge fn fetches apps to see if a given app exists and if does not it creates some record
    * Unfortunately this is more complicated then that. It send a REST HTTP request to supabase [^1]
    * This requires creating a new HTTP connection (so new TCP connection)
    * We have to parse the data we are trying to request into a format supabase can understand (done by supabase.js)
    * Then supabase backend has to parse the HTTP request and convert it into SQL
    * It has to send this SQL into postgress
    * Postgress has to parse this SQL and then get the data and respond
    * Supabase has to parse this SQL response into json
    * This json gets back to our edge fn
    * We have to parse this JSON into JS object
  - Now this update edge fn makes 4 supabase calls
    * The first one is to check if the device version is overwritten (if the device has a forced version linked to it)
     * The second one is to check if the device channel is overwritten (if the device has a forced update channel linked to it)
     * The third one is get the default update channel
     * The last one gets the version (version that the device currently has)
   - After we update the device making another supabase call
   - Then we call R2 to get S3 url
   
This IMPL is very slow. First let's start with how the data is structured.
Both first and second supabase calls (overwrite calls) could be removed if only we have put those overwrites in the `device` table
Then instead of 1 supabase call we would only have 1
Second this default channel changes very rarely (sometimes never) in the lifespan of the app and we check it every call
Third the device update is only sometimes required (if there is any significant change and not just the date)
Lastly it is worth noting that nothing is cached here

To solve this the only way is to restructure the data (Removing both `devices_override` and `channel_devices` tables) and to move close to the data source and handle all logic there.

To move closer to data source we could write a [postgress extension](https://www.postgresql.org/docs/current/spi.html)
As for the language for this extension I see 2 options:

 - [The C programming language](https://en.wikipedia.org/wiki/C_(programming_language))
 - [Rust](https://www.rust-lang.org/) with [pgrx](https://github.com/pgcentralfoundation/pgrx#pgrx)
 
Now I would love to tell you how great of a language C is and how rewriting this endpoint in C could solve all of our performance problems but I am not going to do that. I am no C expert and even a slightest mistake could take down production. Do we really want that?

Now as for rust: I could tell you how great rust is and how it is better then C but the advantages boil down to:
 - Algora contributos (there are a lot of rust project on algora)
 - Safety (C is great until it crashes your db)
 - Big ecosystem of packages
 
Regardless of the language here is how I see this extension:
 - The load balancer stays as is
 - We create a new endpoint (either in rust or TS or C, this can be discussed)
 - This new endpoint is not a edge fn - it maintains a postgress connection and is in place to keep the compatibility. When it gets a request it parses this json request and sends a simple SQL request to this new postgress extensions and then simply parses the response from this postgress extension into json and sends back a response (This should not be an expensive endpoint, most of the compute time will be spend waiting on the network)
  - This new extensions does the following things:
    * It checks if the app exists (can be cached)
    * checks if the device exists and if it does checks if it is overwritten (due to the fact that we are directly in postgress we do not have to use the SQL and can just get the data directly, also this will only be 1 request and it also can be cached)
    * It gets the public channel (cached)
    * It gets the version (cached)
    * It executes most of the logic of the current endpoint returning a record that after parsing to JSON can be sent to the client
    
I mentioned caching a lot. Here is how I see it:
 * Postgress allows for "shared memory" and in rust this is called [PgLwLock](https://docs.rs/pgrx/latest/pgrx/lwlock/struct.PgLwLock.html#).
  * This shared memory is saved between transactions (calls)
  * This memory is mostly going to be read from. Writings will be infrequent
  * Data invalidation can be done using [triggers](https://www.postgresql.org/docs/current/sql-createtrigger.html) thus being super reliable
  * This removes the need for redis and makes the reads way faster. Connecting to redis takes A LOT of steps and a lot of extra overhead
  * Memory should not be that expensive, a more detailed write up of this cache will be done if an impl is created
  * This data can be invalidated (for example data older then 6 hours using some cron trigger)

Unfortunately this extension has a lot of drawbacks:
 * First of all I am not a rust expert (I had written rust before, but it was a while ago) and I am certainly not a C expert
 * Probably the contributors to capgo also are not experts in rust and C
 * This will be complex. Potentially requiring a new hosting (for this proxy) and we will likely have to maintain a new service
 * The deployment of this will be very complicated. Before accepting any solution we would have to rigorously test this to make sure it 100% works and does not behave differently then the current impl
 * Writing tests for this will require a changes to the runner + more tests to make sure all the edge cases are handled properly
 
 Overall I believe experimenting with this is worth it. It could greatly decrease overall costs and make this endpoint way faster.
 I believe it could be worth to conciser creating this and to potentially create a PoC rust (or C) impl and benchmark it to show the true effects of this
     
  [^1]: This will happen multiple times, please not every time I mean `supabase call` I mean this entire lengthy process

**Comment by riderx** [comment]
Thanks a lot for this huge effort on showing what is current and what we could improve.

I do agree with the status, current code is not efficient and do many back and forth.

I believe your proposal is very good, but I would like to share some design decision I have made and want to keep:
- I only use JS, right now changing language could help, but I truly believe we can make things perform enough in JS
- I don't use custom server and will not start now, I'm hard believer in serverless and that something I want to keep, even if the grass looks better outside the field

Outside this point, I already use SQL a lot to make things faster, and I believe we can do more SQL for this endpoint to have less HTTP call and get one result directly from the DB.
I also discovered we can direct connect in Postgres with the serverless function with the pooler
![CleanShot 2023-09-28 at 00 43 51@2x](https://github.com/Cap-go/capgo/assets/4084527/f1eb987b-89a8-432b-9c98-50510d9d2713)

So I believe a tiny step who can already do big change could be to use SQL and SQL function and simplify the table as suggested.

**Comment by MMTE** [comment]
> Currently the updates endpoint is implemented very inefficiently. This was attempted to be fixed by implementing [redis](https://github.com/Cap-go/capgo/blob/d8a70c896b7a44cb289bc0b0fc0de111030b894c/supabase/functions/_utils/redis.ts) but the solution was buggy at best and completely unreliable at worst.
> 
> I think this can be broken down to the following reasons:
> 
> * Redis was never given access to PostgreSQL and it was implemented as a middleware
> * Data was invalidated from an edge function that also was not REALLY that reliable
> * There was a lack of IOS tests done prior to merging redis
> * The underling problem with the slow endpoint was never fixed
> * The entire chain (loadbalancer -> edge fn -> redis ->  supabase backend (supabase HTTP api) -> postgress) was kept
> 
> Here is what currently happens on an update endpoint (with redis disabled):
> 
> * First we trigger the cloudflare load balancer
> * Second we trigger a supabase edge function (or a different provider edge fn)
> * This edge fn fetches apps to see if a given app exists and if does not it creates some record
>   
>   * Unfortunately this is more complicated then that. It send a REST HTTP request to supabase [1](#user-content-fn-1-21f4ce14af6208f487f69c665e38960c)
>   * This requires creating a new HTTP connection (so new TCP connection)
>   * We have to parse the data we are trying to request into a format supabase can understand (done by supabase.js)
>   * Then supabase backend has to parse the HTTP request and convert it into SQL
>   * It has to send this SQL into postgress
>   * Postgress has to parse this SQL and then get the data and respond
>   * Supabase has to parse this SQL response into json
>   * This json gets back to our edge fn
>   * We have to parse this JSON into JS object
> * Now this update edge fn makes 4 supabase calls
>   
>   * The first one is to check if the device version is overwritten (if the device has a forced version linked to it)
>   * The second one is to check if the device channel is overwritten (if the device has a forced update channel linked to it)
>   * The third one is get the default update channel
>   * The last one gets the version (version that the device currently has)
> * After we update the device making another supabase call
> * Then we call R2 to get S3 url
> 
> This IMPL is very slow. First let's start with how the data is structured. Both first and second supabase calls (overwrite calls) could be removed if only we have put those overwrites in the `device` table Then instead of 1 supabase call we would only have 1 Second this default channel changes very rarely (sometimes never) in the lifespan of the app and we check it every call Third the device update is only sometimes required (if there is any significant change and not just the date) Lastly it is worth noting that nothing is cached here
> 
> To solve this the only way is to restructure the data (Removing both `devices_override` and `channel_devices` tables) and to move close to the data source and handle all logic there.
> 
> To move closer to data source we could write a [postgress extension](https://www.postgresql.org/docs/current/spi.html) As for the language for this extension I see 2 options:
> 
> * [The C programming language](https://en.wikipedia.org/wiki/C_(programming_language))
> * [Rust](https://www.rust-lang.org/) with [pgrx](https://github.com/pgcentralfoundation/pgrx#pgrx)
> 
> Now I would love to tell you how great of a language C is and how rewriting this endpoint in C could solve all of our performance problems but I am not going to do that. I am no C expert and even a slightest mistake could take down production. Do we really want that?
> 
> Now as for rust: I could tell you how great rust is and how it is better then C but the advantages boil down to:
> 
> * Algora contributos (there are a lot of rust project on algora)
> * Safety (C is great until it crashes your db)
> * Big ecosystem of packages
> 
> Regardless of the language here is how I see this extension:
> 
> * The load balancer stays as is
> * We create a new endpoint (either in rust or TS or C, this can be discussed)
> * This new endpoint is not a edge fn - it maintains a postgress connection and is in place to keep the compatibility. When it gets a request it parses this json request and sends a simple SQL request to this new postgress extensions and then simply parses the response from this postgress extension into json and sends back a response (This should not be an expensive endpoint, most of the compute time will be spend waiting on the network)
> * This new extensions does the following things:
>   
>   * It checks if the app exists (can be cached)
>   * checks if the device exists and if it does checks if it is overwritten (due to the fact that we are directly in postgress we do not have to use the SQL and can just get the data directly, also this will only be 1 request and it also can be cached)
>   * It gets the public channel (cached)
>   * It gets the version (cached)
>   * It executes most of the logic of the current endpoint returning a record that after parsing to JSON can be sent to the client
> 
> I mentioned caching a lot. Here is how I see it:
> 
> * Postgress allows for "shared memory" and in rust this is called [PgLwLock](https://docs.rs/pgrx/latest/pgrx/lwlock/struct.PgLwLock.html#).
> * This shared memory is saved between transactions (calls)
> * This memory is mostly going to be read from. Writings will be infrequent
> * Data invalidation can be done using [triggers](https://www.postgresql.org/docs/current/sql-createtrigger.html) thus being super reliable
> * This removes the need for redis and makes the reads way faster. Connecting to redis takes A LOT of steps and a lot of extra overhead
> * Memory should not be that expensive, a more detailed write up of this cache will be done if an impl is created
> * This data can be invalidated (for example data older then 6 hours using some cron trigger)
> 
> Unfortunately this extension has a lot of drawbacks:
> 
> * First of all I am not a rust expert (I had written rust before, but it was a while ago) and I am certainly not a C expert
> * Probably the contributors to capgo also are not experts in rust and C
> * This will be complex. Potentially requiring a new hosting (for this proxy) and we will likely have to maintain a new service
> * The deployment of this will be very complicated. Before accepting any solution we would have to rigorously test this to make sure it 100% works and does not behave differently then the current impl
> * Writing tests for this will require a changes to the runner + more tests to make sure all the edge cases are handled properly
> 
> Overall I believe experimenting with this is worth it. It could greatly decrease overall costs and make this endpoint way faster. I believe it could be worth to conciser creating this and to potentially create a PoC rust (or C) impl and benchmark it to show the true effects of this
> 
> ## Footnotes
> 1. This will happen multiple times, please not every time I mean `supabase call` I mean this entire lengthy process [↩](#user-content-fnref-1-21f4ce14af6208f487f69c665e38960c)

Hello.

Thank you very much for your explanation.
I'm trying to think more about this and since I still don't know where the bottleneck is, it's hard for me to talk about it, but for now, I have a few things in mind:

1- How much will this help to update the programs? Do we have statistics or numbers of results that we can see the speed changes and make decisions based on? Supposedly, with these changes and the development of the plugin for Postegras, we were now responding to 33,000 requests per second. What will be our approach for 1 million requests per second in the future?
PS: yesterday @riderx mentioned in Discord that this process is still under 900ms which is not bad.
2- Apart from the discussion of how much we can increase the speed of updating, how important is this speed increase in terms of business or user experience in this request? Is this speed boost necessary or nice to have?
3- I don't want to complicate things or use strange patterns, but I think that despite increasing the speed, which will be a good option, we should also look at the update action. For example, currently, an update request is given then a TCP connection is established and the client is waiting for a response, while if we use a web socket, we can respond to the request and give the final result in another request. I don't know how useful this can be. on this topic I recently found [pushpin](https://pushpin.org/) which seems interesting to me. 
4- I think using a queue service would be helpful in this situation because we have high requests per second. I guess currently It would solve two issues. one is domain failover service which @riderx mentioned has an expensive billing. since the queue system gets requests from the user and keeps them in the queue and then it processes them in order. also, most queue systems are scaleable so in case of high requests it can be scaled. I don't know if it would help domain failover or not but I'm positive about that. I guess the second problem that a queue can fix is the load we have on supabase function calls which engage PostgreSQL. As the request gets higher on business growth it doesn't matter when but someday request volume will defeat the performance side so then we need to scale either horizontally and add servers or vertically with replications.
In the end, I think supabase support team can help with this issue too. I guess they must have customers with the same problems and experience. maybe they can share some of their viewpoints. 

Please tell me if you have any points about these things.

**Comment by WcaleNieWolny** [comment]
> Do we have statistics or numbers of results that we can see the speed changes and make decisions based on

No we do not, we do not even have a PoC and it is impossible to know how much will this improve the code speed until we write a PoC

> What will be our approach for 1 million requests per second in the future?
Make this plugin more efficient? Even if we ever switch to scaling horizontally having an efficient endpoint would allays be better then having a slow one

> Apart from the discussion of how much we can increase the speed of updating, how important is this speed increase in terms of business or user experience in this request

This plugin is really more about removing overhead. Removing overhead is removing CPU time and removing CPU time equals less cost. The speed is just a nice byproduct

> domain failover
We can just disable this failover if it costs so much

> so then we need to scale either horizontally
The thing about supabase is that it is REALLY not supposed to scale horizontally. If we ever want to scale horizontally we would have to undergo a MASSIVE migration and this is not the topic for today in my opinion

> I think supabase support team can help with this issue too
I agree, we can ask them

**Comment by riderx** [comment]
I have currently made 2 big change who did reduce a lot the charge of the db:
- I have commented the store_apps upsert ( and need to migrate it to clickhouse too) 
- I have move out, devices and stats table to clickhouse

So right now it seems Supabase is handling way more of the load.
I still believe making less call to Postgres is a good way to improve the latency

**Comment by riderx** [comment]
I think the next move is this:
https://github.com/Cap-go/capgo/issues/460
So closing in favor of D1
